python/dgl/graphbolt/utils/sample_utils.py

```diff
@@ -264,6 +339,71 @@ def unique_and_compact_node_pairs(
     # Return singleton for a homogeneous graph.
     if is_homogeneous:
         compacted_node_pairs = list(compacted_node_pairs.values())[0]
-        unique_nodes = list(unique_nodes.values())[0]
+        cat_seeds = list(cat_seeds.values())[0]
+
+    return cat_seeds, compacted_node_pairs
+
+
+def not_deduplication_compact_dst(
+    dst: torch.Tensor, duplicated_dst: torch.Tensor
+):
+    """
+    Compact dst without deduplication.
 
-    return unique_nodes, compacted_node_pairs
+    Parameters
+    ----------
+    dst: torch.Tensor
+        The original dst that need to be compacted.
+    duplicated_dst: torch.Tensor
+        Duplicated_dst record the nodes that generate dst in order. Its index
+        will be used to help generate the compacted dst.
+
+    Returns
+    -------
+    torch.Tenor
+        Compacted dst without deduplication.
+    """
+    # Record the number of consecutive identical values.
+    dst_constinuous = [1 for _ in duplicated_dst]
```

Comment:
as this function will be called in critical path, please pay attention on performance.

Reply:
could you describe the algo and implementation details to help us have a basic and clear understanding.

Reply:
## Step 1: Initialize Supporting Structures
    
`seeds_constinuous`: A list tracking consecutive occurrences of elements in seeds.
`dst_compact`: An empty list, later to contain mapped subscripts corresponding to elements in dst.
`compacted_dst_num`: An empty dictionary, which will eventually record the count of appearances of each element in the compacted destination tensor.

**Objective**: Ensure we have structures to store subsequent calculations and mappings.

## Step 2: Establish Initial Mappings

**Action**: Employ a while loop to traverse through `dst` and `seeds` in reverse order.
**Goal**: Populate `dst_compact` and `compacted_dst_num` by establishing a preliminary relationship between corresponding elements in `dst` and `seeds`. Specifically:

- `dst_compact` will contain subscripts pointing to related elements between `dst` and `seeds`.

- `compacted_dst_num` keeps track of how often each subscript appears in `dst_compact`.

**Objective**: Create a rudimentary mapping between dst and the compacted destination tensor, albeit requiring further refinement.

## Step 3: Refine Mappings

**Action**: Traverse dst_compact in reverse once again.
**Goal**: Correct the initial mappings by:

- Checking for sequential duplicates in `dst`.

- Adjusting the mapped values in `dst_compact` to ensure they correctly correspond with the `dst` and `seeds` by calculating match_times and the number of consecutive seeds with same value.

**Objective**: Ensure that each value in `dst` is mapped to a correct, and distinctly corresponding, seed subscript, even when dealing with consecutive identical values.

## Step 4: Finalize and Return Result

**Action**: Convert `dst_compact` into a torch tensor.
**Return**: This torch tensor, which is the refined, compacted destination tensor, `compacted_dst`.

**Objective**: Output a final, compacted tensor that accurately represents dst in a form that’s mapped via subscripts to seeds.